packages/rich-list-core/src/RichListCore.ts

```diff
@@ -0,0 +1,174 @@
+import { ApiClient, BigNumber, blockchain as defid } from '@defichain/jellyfish-api-core'
+import { QueueClient } from './lib/Queue'
+import { PersistentStorage } from './lib/PersistentStorage'
+import { AddressParser } from './controller/AddressParser'
+import { NetworkName } from '@defichain/jellyfish-network'
+import { RichListItem } from '@defichain/rich-list-api-client'
+import { AccountAmount } from 'packages/jellyfish-api-core/src/category/account'
+import { ActiveAddressAccountAmount } from './controller/AddressParser/ActiveAddressAccountAmount'
+import { PersistentLinkedList } from './lib/PersistentLinkedList'
+
+const DEFAULT_RICH_LIST_LENGTH = 1000
+
+export class RichListCore {
+  isCatchingUp = false
+  richListLength = DEFAULT_RICH_LIST_LENGTH
+  readonly addressParser: AddressParser
+
+  constructor (
+    private readonly network: NetworkName,
+    private readonly apiClient: ApiClient,
+    private readonly existingRichList: PersistentStorage<number, RichListItem[]>,
+    private readonly queueClient: QueueClient<string>,
+    private readonly crawledBlocks: PersistentLinkedList<BlockDroppedOutAddresses>
+  ) {
+    this.addressParser = new AddressParser(apiClient, network)
+  }
+
+  setRichListLength (length: number): void {
+    this.richListLength = length
+  }
+
+  resume (): void {
+    if (this.isCatchingUp) {
+      return
+    }
+    this.isCatchingUp = true
+    void this._catchUp()
+  }
+
+  async _catchUp (): Promise<void> {
+    const nextBlockHeight = await this.crawledBlocks.size()
+    const nextBlock = await this._getBlock(nextBlockHeight)
+
+    if (nextBlock === undefined) {
+      this.isCatchingUp = false
+      return
+    }
+
+    const lastHashed = await this.crawledBlocks.getLast()
+    if (lastHashed !== undefined && lastHashed.hash !== nextBlock.previousblockhash) {
+      // TODO(@ivan-zynesis): invalidate()
+    } else {
+      const queue = await this.queueClient.createQueueIfNotExist(this._queueName(), 'LIFO')
+
+      const addresses: string[] = []
+      for (const tx of nextBlock.tx) {
+        const _addresses = await this.addressParser.parse(tx)
+        for (const a of _addresses) {
+          addresses.push(a)
+          await queue.push(a)
+        }
+      }
+      await this.crawledBlocks.append({
+        hash: nextBlock.hash,
+        richListDroppedOut: addresses
+      })
+    }
+
+    return await this._catchUp()
+  }
```

Comment:
```suggestion
      await this.crawledBlocks.append({
        hash: nextBlock.hash,
        richListDroppedOut: await this._getAddresses(nextBlock)
      })
    }

    return await this._catchUp()
  }
  
  private async _getAddresses(block: defid.Block<defid.Transaction>): Promise<string[]> {
    const queue = await this.queueClient.createQueueIfNotExist(this._queueName(), 'LIFO')
    const addresses: string[] = []
    for (const tx of block.tx) {
      const _addresses = await this.addressParser.parse(tx)
      for (const address of _addresses) {
        addresses.push(address)
        await queue.push(address)
      }
    }
    return addresses
  }
```
This one is more of a personal observation, looking at the overall method growing in size we may want to try to cut down the size and move some logic out into a separate method. So this is mostly just a personal suggestion than a blocker.

Reply:
Adopt the suggestion, also fixed a dumb mistake, the `droppedOutAddress` was a silly mistake, they are only computed after compared with previous rich list. 😓